pyanaconda/modules/payloads/payload/live_image/installation_progress.py

```diff
@@ -0,0 +1,123 @@
+#
+# Copyright (C) 2021  Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+import os
+import time
+
+from blivet.size import Size
+
+from pyanaconda.anaconda_loggers import get_module_logger
+from pyanaconda.core.constants import THREAD_LIVE_PROGRESS
+from pyanaconda.core.i18n import _
+from pyanaconda.core.util import join_paths
+from pyanaconda.modules.common.constants.objects import DEVICE_TREE
+from pyanaconda.modules.common.constants.services import STORAGE
+from pyanaconda.modules.common.task.cancellable import Cancellable
+from pyanaconda.threading import threadMgr, AnacondaThread
+
+log = get_module_logger(__name__)
+
+__all__ = ["InstallationProgress"]
+
+
+class InstallationProgress(Cancellable):
+    """Progress monitor of the image installation."""
+
+    def __init__(self, sysroot, installation_size, callback):
+        """Create a new installation progress.
+
+        :param sysroot: a path to the system root
+        :param installation_size: a size of the installed payload
+        :param callback: a function for the progress reporting
+        """
+        super().__init__()
+        self._sysroot = sysroot
+        self._callback = callback
+        self._installation_size = installation_size
+        self._thread_name = THREAD_LIVE_PROGRESS
+
+    def __enter__(self):
+        """Start to monitor the progress."""
+        # Start the thread.
+        threadMgr.add(AnacondaThread(
+            name=self._thread_name,
+            target=self._monitor_progress
+        ))
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Stop to monitor the progress."""
+        # Cancel the progress reporting.
+        self.cancel()
+
+        # Wait for the thread to finish.
+        threadMgr.wait(self._thread_name)
+
+    def _monitor_progress(self):
+        """Monitor the amount of disk space used on the target and source."""
+        log.debug("Installing %s.", Size(self._installation_size))
+
+        # Force write everything to disk.
+        self._callback(_("Synchronizing writes to disk"))
+        os.sync()
```

Comment:
It's seems a bit weird to me to assure data is synced out during image install from progress monitoring thread/callback. If it was really just progress monitoring then it should not have side effect like this.

I would suggest either splitting this to a separate data sync callback/thread or renaming the combined callback/thread to something like "synchronization_and_monitoring" to make it evident it's not just read only monitoring without any side effects.

Reply:
The monitoring is based on `os.statvfs` results. The synchronization ensures that the starting size bellow is up to date. If we download the image and don't synchronize the writes, the progress reporting doesn't return correct values, because the size of the downloaded image messes with the currently used size. Therefore, the synchronization is not a side effect, but part of the monitoring.

Reply:
Just how costly is `os.sync()` ? I assume the writes need to happen at some time anyway, and we don't call this often enough to cause more writes than needed?

Reply:
Yep, the writes will happen anyway.

Reply:
Oh, the `os.sync()` call is there to keep the `os.statvfs` values up to date for progress monitoring purposes. In that case I'm fine with it being called in the monitoring thread, but please note that also in in the code that it's there for monitoring not data integrity purposes. Thanks! :)